pkg/proxy/middleware.go

```diff
@@ -0,0 +1,253 @@
+package proxy
+
+import (
+	"bytes"
+	"errors"
+	"io"
+	"math/rand"
+	"net/http"
+	"time"
+)
+
+var _ http.Handler = &RetryMiddleware{}
+
+type RetryMiddleware struct {
+	nextHandler      http.Handler
+	maxRetries       int
+	retryStatusCodes map[int]struct{}
+}
+
+// NewRetryMiddleware creates a new HTTP middleware that adds retry functionality.
+// It takes the maximum number of retries, the next handler in the middleware chain,
+// and an optional list of retryable status codes.
+// If the maximum number of retries is 0, it returns the next handler without adding any retries.
+// If the list of retryable status codes is empty, it uses a default set of status codes (http.StatusBadGateway, http.StatusServiceUnavailable, http.StatusGatewayTimeout).
+// The function creates a RetryMiddleware struct with the given parameters and returns it as an http.Handler.
+func NewRetryMiddleware(maxRetries int, other http.Handler, optRetryStatusCodes ...int) http.Handler {
```

Comment:
We need to limit the max bytes read.  See https://cwe.mitre.org/data/definitions/400.html example 6

Reply:
I can see this being useful.